web/client_service_indexeddb.ts

```diff
@@ -0,0 +1,111 @@
+import Dexie, {Table} from 'dexie';
+import {KEYUTIL, b64utohex} from 'jsrsasign';
+import {Properties, CLIENT_PROPERTIES_FIELD} from './common/client_config';
+import {ClientService} from './client_service';
+import {isNonEmptyString, isInteger} from './common/polyfill/is';
+
+const KEYSTORE_DATABASE_NAME = 'scalar';
+
+class ScalarIndexedDb extends Dexie {
+  keystore!: Table<{id: string; key: CryptoKey}, string>;
+  certstore!: Table<{id: string; cert: string}, string>;
+
+  constructor() {
+    super(KEYSTORE_DATABASE_NAME);
+    this.version(1).stores({keystore: 'id', certstore: 'id'});
+  }
+}
+
+export class ClientServiceWithIndexedDb extends ClientService {
+  static async create(properties: Properties) {
```

Comment:
Here is the main logic of ClientServiceWithIndexedDb.

If a key (certificate) can be found in the properties, then store it into indexedDB, otherwise, ClientServiceWithIndexedDb tries to load it and fill into the properties.